protected/humhub/components/ModuleManager.php

```diff
@@ -146,79 +121,73 @@ public function registerBulk(array $configs)
     {
-        $filename = $basePath . '/config.php';
+        $filename = $basePath . DIRECTORY_SEPARATOR . ModuleAutoLoader::CONFIGURATION_FILE;
         if ($config === null && is_file($filename)) {
             $config = include $filename;
         }
 
         // Check mandatory config options
-        if (!isset($config['class']) || !isset($config['id'])) {
-            throw new InvalidConfigException('Module configuration requires an id and class attribute: ' . $basePath);
-        }
+        $moduleInfo = ModuleInfo::instantiate($config, $basePath);
 
-        $isCoreModule = (isset($config['isCoreModule']) && $config['isCoreModule']);
-        $isInstallerModule = (isset($config['isInstallerModule']) && $config['isInstallerModule']);
+        // ToDo: here we keep the *last* loaded module with a given id.
+        //       But when it comes to aliases further down below, then we consider only the *first" one
+        $this->modules[$moduleInfo->moduleId] = $moduleInfo;
 
-        $this->modules[$config['id']] = $config['class'];
-
-        if (isset($config['namespace'])) {
-            Yii::setAlias('@' . str_replace('\\', '/', $config['namespace']), $basePath);
+        if ($moduleInfo->namespace) {
+            Yii::setAlias('@' . str_replace('\\', '/', $moduleInfo->namespace), $basePath);
         }
 
         // Check if alias is not in use yet (e.g. don't register "web" module alias)
-        if (Yii::getAlias('@' . $config['id'], false) === false) {
-            Yii::setAlias('@' . $config['id'], $basePath);
+        if (Yii::getAlias('@' . $moduleInfo->moduleId, false) === false) {
+            Yii::setAlias('@' . $moduleInfo->moduleId, $basePath);
         }
 
-        if (isset($config['aliases']) && is_array($config['aliases'])) {
-            foreach ($config['aliases'] as $name => $value) {
-                Yii::setAlias($name, $value);
-            }
+        array_walk(
+            $moduleInfo->aliases,
+            static fn($value, $name) => Yii::setAlias($name, $value)
+        );
+
+        if (!Yii::$app->params['installed'] && $moduleInfo->isInstallerModule) {
+            $moduleInfo->isRegistered = true;
         }
+        elseif (Yii::$app->params['databaseInstalled'] && Yii::$app->params['installed'])
+        {
+            $registration = ModuleEnabled::getRegisteredModuleInfo($moduleInfo->moduleId);
 
-        if (!Yii::$app->params['installed'] && $isInstallerModule) {
-            $this->enabledModules[] = $config['id'];
+            $moduleInfo->isRegistered = $registration !== null;
+            $moduleInfo->isPaused     |= ($moduleInfo->isRegistered && $registration->isPaused);
         }
 
         // Not enabled and no core/installer module
-        if (!$isCoreModule && !in_array($config['id'], $this->enabledModules)) {
+        if (!$moduleInfo->isCoreModule && !$moduleInfo->getIsActive()) {
             return;
         }
 
-        // Handle Submodules
```

Comment:
Has submodule handler been moved somewhere else for better use?

Reply:
Since the `ModuleInfo->modules` property is an already-initialized array, it is not required to do the check in line 189 anymore.

Whether or not the module is a core module is also determined during instantiation of `ModuleInfo` and stored as a property thereof.